### DISCUSSION TITLE
关于通过日志内存的信息分析内存使用情况的概要设计

### DISCUSSION THREAD

**Discussion by oraclebird** [oceanbase/obdiag#453] [Ideas]
# 需求背景
>需求来源
observer日志中有内存的信息打印，obdiag支持一下从日志中分析内存使用情况 
#120
>需求梳理
前提需求：从observer中解析出内存信息并存入指定文件中
实际需求：内存信息趋势展示，从内存信息中找出异常的内存信息，输出结果

# 功能性设计

### 命令行设计
第一种方案：新增memory二级功能
分析内存信息：
obdiag analyze memory 
obdiag analyze memory --ctx_name = ctx_name
obdiag analyze memory --ctx_name = ctx_name --mod_name = mod_name
第二种方案：在log的基础上添加memory分析
obdiag analyze log --type=memory
obdiag analyze log --type=memory --ctx_name = ctx_name
obdiag analyze log --type=memory --ctx_name = ctx_name --mod_name = mod_name
其他的选项如开始时间结束时间可参考obdiag analyze log

### 可视化设计
可视化打印主要安装第三方可视化插件，目前OB还没有引入过可视化插件，可视化插件包也比较大，可能会对obdiag后续的安装部署有影响，目前这部分要进行讨论。目前我这边选了plotly

# 非功能性设计

## 安全
> 不涉及

## 文档

# 影响评估

## 兼容性评估
 兼容OB3.x和OB4.x版本

## 风险评估
> 如果分析的时间过长，可能要下载较多的observer.log，可能会造成一定的网络拥塞和CPU消耗

## 性能评估
> 评估该功能在性能上的影响。

## 测试分析
> 对于ob3.x和ob4.x各个小版本均要进行测试

# 工作量评估

## 开发计划
> 开发联调目标和计划。

| 任务编号 | 任务名称 | 负责人 | 开始日期 | 结束日期 | 计划用时|
|---------|---------|-------|---------|---------|----|
| 1       | 设计文档 | xx   | 2024-09-25 | 2024-09-27 | 3d|
| 2       | 编码实现 |xx   | 2024-10-8 | 2024-10-15 | 7d|
| 3       | 自测 | xx   | 2024-10-16 | 2024-10-20 |5d|
| 4       | 提测/验收 | xx   | 2024-10-21 | 2024-10-30 | 10d|

**Comment by Teingi** [comment]
9月27号评审结论：
- 评审通过
- 按照方案一执行
- 结果打印部分需要评估引入包后编译出来的二进制大小，如果编译后新增超过5MB，则不允许引入
- 该功能是否跟随2.5.0迭代不做强制要求，质量第一位
